src/content/blog/2025-01-02-python-shorts-pivoting-in-pandas.mdx

```diff
@@ -0,0 +1,200 @@
+
+We can create an appropriately-shaped dataframe from our hierarchically-indexed series by resetting its index, which drops both levels of the index (BLT_DECADE and SUBTYPE) into their own columns of categorical values. Now that we've got a proper long dataframe, we can pivot that to get the wide dataframe we want.
+
+```python
+hui_dataframe.groupby(["BLT_DECADE", "SUBTYPE"])["UNIT_COUNT"].sum()
+             .reset_index()
+             .pivot(index="BLT_DECADE", columns="SUBTYPE", values="UNIT_COUNT")
+```
+
+<Image
+  src={StackedDataframe}
+  loading="eager"
+  alt="The pivoted DataFrame has the same values as the unstacked dataframe but a different column order."
+/>
+
+## Melt Me Down
+
+We've been focused on going from long to wide, but Pandas also has tools for going from wide to long. It's [`melt`](https://pandas.pydata.org/docs/user_guide/reshaping.html#melt-and-wide-to-long) function allows you to automatically do the transformation I did by hand to the table at the beginning- choose a column (or columns) as ID variables and stack the rest of the value columns into two "variable" and "value" columns.
+
+## But Can It Do Pivot Tables?
+
+Ask someone what makes them an Excel expert, and chances are you'll hear "pivot tables" in their explanation. This combines the three steps we've done to the HUI data so far: grouping the data by categorical columns, performing some aggregation function on a numerical column, and pivoting the results into an easy-to-parse two-dimensional table.
+
+Because this specific case of pivoting is so prevalent, Pandas has its own built-in version of it. This single function replaces the `groupby`, `sum`, and `pivot`/`unstack` functions we've been chaining together.
+
+```python
+pd.pivot_table(hui_dataframe, values="UNIT_COUNT", index="BLT_DECADE", columns="SUBTYPE", aggfunc="sum")
+```
+
+This again gives us the table we're familiar with by this point:
+
+<Image
+  src={PivotTable}
+  loading="eager"
+  alt="The resulting DataFrame looks similar to the previous unstacked and pivoted DataFrames"
+/>
+
+## Coming Full Circle
+
+So why didn't we just lead with `pivot_table`? First, this is a more limited and specific case of pivoting for when you want to perform one of the common aggregator functions (Which functions are those? As usual, [StackOverflow to the rescue](https://stackoverflow.com/questions/53943319/what-are-all-pandas-agg-functions)! Basically, it'll try anything you throw at it, but there are some common ones.).
+
+`pivot` and `unstack`/`stack` will apply to other situations where an earlier analysis step gives you a hierarchically indexed data structure or creates columns with categorical group identifier. It can also be useful to reshape the results of joins into the wide format expected by GIS software.
```

Comment:
I think this needs the a or the s.

```suggestion
`pivot` and `unstack`/`stack` will apply to other situations where an earlier analysis step gives you a hierarchically indexed data structure or creates columns with a categorical group identifiers. It can also be useful to reshape the results of joins into the wide format expected by GIS software.
```

Reply:
Good catch, thanks.